### PR TITLE
Fix redirection of stdout and stderr to suite logs

### LINKED ISSUE #2411: suite logging: restart can write to old log
The issue is caused by the redirection in `daemonize()` which latches onto the file handler before the suite logs have been initalised and thus before the logs have been rolled. This means that output after restart can appear in both the log from the previous run.

Initialising the suite logs before the `daemonize()` function fixes this. I need to test this more thoroughly before putting up a pull request.

__Test to detect this issue:__

_tests/logging/02-duplicates.t_
```bash
#!/bin/bash
# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
# Copyright (C) 2008-2017 NIWA
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#-------------------------------------------------------------------------------
. $(dirname $0)/test_header
#-------------------------------------------------------------------------------
set_test_number 3
#-------------------------------------------------------------------------------
install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
gcylc ${SUITE_NAME} &
suite_run_ok "${TEST_NAME_BASE}-run" cylc run "${SUITE_NAME}"
while cylc ping "${SUITE_NAME}" 2>/dev/null; do
    sleep 1
done
sleep 8
suite_run_ok "${TEST_NAME_BASE}-restart" cylc restart "${SUITE_NAME}"
while cylc ping "${SUITE_NAME}" 2>/dev/null; do
    sleep 1
done
if [[ -e "${SUITE_RUN_DIR}/work/2/pub/test-succeeded" ]]; then
    ok "${TEST_NAME_BASE}-check"
else
    fail "${TEST_NAME_BASE}-check"
    echo 'OUT - Duplicated Entries:' >&2
    cat "${SUITE_RUN_DIR}/work/2/pub/out-duplication" >&2
    echo 'ERR - Duplicated Entries:' >&2
    cat "${SUITE_RUN_DIR}/work/2/pub/err-duplication" >&2
    echo 'LOG - Duplicated Entries:' >&2
    cat "${SUITE_RUN_DIR}/work/2/pub/log-duplication" >&2
fi

#-------------------------------------------------------------------------------
#purge_suite "${SUITE_NAME}"
exit
```

_tests/logging/02-duplicates/suite.rc_

```ini
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[dependencies]]
        [[[R1/1]]]
            graph = """
                foo:fail => bar
                foo & bar => restart
            """
        [[[R1/2]]]
            graph = """
                restart[-P1] => foo
                foo:fail => bar
                foo & bar => pub
            """
[runtime]
    [[foo]]
        script = false
    [[bar]]
        script = """
cylc reset "${CYLC_SUITE_NAME}" "foo.${CYLC_TASK_CYCLE_POINT}" -s succeeded
"""
    [[restart]]
        script = """
cylc stop "${CYLC_SUITE_NAME}"
"""
    [[pub]]
        script = """
# Extract timestamp lines from logs
for file in $(find "${CYLC_SUITE_RUN_DIR}/log/suite/" -name '*.*'); do
    if $(grep '.*-.*-.*' "${file}"); then
        grep '.*-.*-.*' "${file}" | sort -u > $(basename $file)
    else
        touch $(basename $file)
    fi
done

# Write out duplicate entries to *-duplication files.
sort $(find . -name 'out*') | uniq -d > out-duplication
sort $(find . -name 'err*') | uniq -d > err-duplication
sort $(find . -name 'log*') | uniq -d > log-duplication

# Fail if any of these files contain any content.
if [[ -s out-duplication || -s err-duplication || -s log-duplication ]]; then
    touch 'test-failed'
else
    touch 'test-succeeded'
fi
"""
```

### PR DESCRIPTION
**Pull request by oliver-sanders**
Close #2411 

Changes:
- When the suite is running in detached mode stdout and stderr redirected to log/out and log/err respectively. When the logs rotate these streams are re-directed again so as to pick up the new file.
- Logs are only re-directed to stdout/stderr when running in non-detatched mode

### CODE REVIEW

**Comment by oliver-sanders** [comment]
The test can be cleaned up a little, it'll do for now. I'll add a further test for sequential timestamps next week, these changes can go into a future cylc release.

File: lib/cylc/scheduler.py [hunk 1]
```diff
@@ -217,14 +217,17 @@ def start(self):
             self.suite_db_mgr.restart_upgrade()
 
         try:
-            if not self.options.no_detach and not cylc.flags.debug:
+            detatch = not (self.options.no_detach or cylc.flags.debug)
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: detatch spelled wrong, should be "detach" (although it's not a bug!)

**Comment by hjoliver** [comment]
The code looks fine, but the test always passes for me on master - presumably that should not happen?

**Comment by oliver-sanders** [comment]
This fixes #2411 so the test should pass.

**Comment by oliver-sanders** [comment]
Missed a `-q` from the `grep` command, test should now fail on master.

**Comment by hjoliver** [comment]
All good, thanks.  @matthewrmshin - note I've updated the change log for 7.5.0, and will attempt to make the release Wednesday night. I'm in a workshop all day tomorrow, and on leave Thursday and Friday.  Worst case, release on the weekend...

**Pull request merged**